packages/cells/src/widget.ts

```diff
@@ -832,6 +845,175 @@ export namespace CodeCell {
+    event.stopPropagation();
+    if (event.proposedAction === 'none') {
+      event.dropAction = 'none';
+      return;
+    }
+
+    const thunks = event.mimeData.getData(ATTACHMENTS_MIME);
+    const promises: Promise<Contents.IModel>[] = [];
+    thunks.forEach((fn: Function) => {
+      promises.push(fn());
+    });
+    Promise.all(promises).then(models => {
+      models.forEach(model => {
+        if (model.type === 'file' && model.format === 'base64') {
+          this.model.attachments.set(model.name, {
+            [model.mimetype]: model.content
+          });
+          this._updateCellSourceWithAttachment(model.name);
+        }
+      });
+    });
+  }
+
+  /**
+   * Handle `after-attach` messages for the widget.
+   */
+  protected onAfterAttach(msg: Message): void {
+    super.onAfterAttach(msg);
+    let node = this.node;
+    node.addEventListener('p-dragover', this);
+    node.addEventListener('p-drop', this);
+    node.addEventListener('dragenter', this);
+    node.addEventListener('dragover', this);
+    node.addEventListener('drop', this);
+    node.addEventListener('paste', this);
+  }
+
+  /**
+   * A message handler invoked on a `'before-detach'`
+   * message
+   */
+  protected onBeforeDetach(msg: Message): void {
+    let node = this.node;
+    node.removeEventListener('drop', this);
+    node.removeEventListener('dragover', this);
+    node.removeEventListener('dragenter', this);
+    node.removeEventListener('paste', this);
+    node.removeEventListener('p-dragover', this);
+    node.removeEventListener('p-drop', this);
+  }
+
+  /**
+   * Attaches all DataTransferItems (obtained from
+   * clipboard or native drop events) to the cell
+   */
+  private _attachFiles(items: DataTransferItemList) {
+    for (let i = 0; i < items.length; i++) {
+      const item = items[i];
+      if (item.kind === 'file') {
+        const blob = item.getAsFile();
+        this._attachFile(blob);
+      }
+    }
+  }
+
+  /**
+   * Takes in a file object and adds it to
+   * the cell attachments
+   */
+  private _attachFile(blob: File) {
+    const reader = new FileReader();
+    reader.readAsDataURL(blob);
+    reader.onload = evt => {
+      const { pathname } = URLExt.parse(reader.result as string);
+      const [mimeType, encodedData] = pathname.split(';');
+      const data = encodedData.split(',')[1];
+      const bundle: nbformat.IMimeBundle = { [mimeType]: data };
+      this.model.attachments.set(blob.name, bundle);
+      this._updateCellSourceWithAttachment(blob.name);
+    };
+    reader.onerror = evt => {
+      console.error(`Failed to attach ${blob.name}` + evt);
+    };
+  }
+
+  /**
+   * Appends the text ![attachment](attachment: <attachmentName>)
+   * to the cell source
+   */
+  private _updateCellSourceWithAttachment(attachmentName: string) {
+    const textToBeAppended = `![attachment](attachment:${attachmentName})`;
+    this.model.value.text = `${this.model.value.text}\n${textToBeAppended}`;
```

Comment:
I think this code from https://github.com/agoose77/jupyterlab-attachments can be used to advantage here (modified slightly):

```ts
/**
 * Insert Markdown code to embed image attachment.
 */
insertImageFromAttachment(attachmentName: string) {
    // Markdown template string to insert image
    const markdown = `![${attachmentName}](attachment:${attachmentName})`;
    const value = this.model.value;
    value.insert(value.text.length, markdown);
}
```

Reply:
Oh I didn't know this existed. Thanks!

Reply:
Taken care of this